Add implementation and tests for Stacks

wgj opened the pull request:
Adds Gradle stuff

Per CTCI: 
> [Stuff in README.md] is important. If you don't feel very, very comfortable with each of thes data structures and algorithms listed, practice implementing them from scratch.

@sabeechen, is this a good way to share and review progress? Are the tests "good" and "professional"?

wgj committed:
Add implementation and tests for Stacks

sabeechen commented on src/main/java/io/wgj/ctci/Stack.java:
```diff
@@ -0,0 +1,21 @@
+    int[] stack;
+    int position;
+
+    public Stack(){
```
Pedantic: needs whitespace before the "{" here and in pop()

sabeechen commented on src/main/java/io/wgj/ctci/Stack.java:
```diff
@@ -0,0 +1,21 @@
+package io.wgj.ctci;
+
```
Recommend having comments describing how the stack is supposed to be used.  For example, is it supposed to max out at 10 elements?

wgj replied:
Fixed in cd6dc83cc0daf94827e214ba2a7c26f33c9aef00

sabeechen commented on src/main/java/io/wgj/ctci/Stack.java:
```diff
@@ -0,0 +1,21 @@
+package io.wgj.ctci;
+
+public class Stack {
+    int[] stack;
```
I'm not sure if this is *supposed* to only handle 10 elements max, but if not I'd recommend using an ArrayList for simplicity.  Then position is always stack.size() and it grows dynamically.
BONUS SUGGESTION: Implementing a [Dynamic Array](https://en.wikipedia.org/wiki/Dynamic_array) in here without using the standard librarys would be a great exercise.

wgj replied:
Fixed in cd6dc83cc0daf94827e214ba2a7c26f33c9aef00

sabeechen commented on src/main/java/io/wgj/ctci/Stack.java:
```diff
@@ -0,0 +1,21 @@
+        position--;
+        return stack[position];
+    }
+}
```
IMHO this class needs a peek() method, a size() method, and *maybe* a maxSize() method to be usable in practice unless you're intending to abstract away from those details.

wgj replied:
What should `maxSize()` return?

sabeechen replied:
The maxSize of the stack.  10 in current implementation.

wgj replied:
Fixed in cd6dc83cc0daf94827e214ba2a7c26f33c9aef00

sabeechen commented on src/test/java/io/wgj/ctci/StackTest.java:
```diff
@@ -0,0 +1,50 @@
+    @Test(expected = ArrayIndexOutOfBoundsException.class)
+    public void testStackUnderFlow() {
+        Stack s = new Stack();
+        s.pop();
```
Pedantic: Could be simplified to just "new Stack().pop()"

sabeechen commented on src/test/java/io/wgj/ctci/StackTest.java:
```diff
@@ -0,0 +1,50 @@
+    @Test(expected = ArrayIndexOutOfBoundsException.class)
+    public void testStackOverFlow() {
+        Stack s = new Stack();
+        for (int i = 0; i < 11; i++) {
```
I think this suffers from the problem that if the exception were thrown for one of the first 10 calls for push(), it would still pass.  I think what you want it to make 10 push calls, then make an [@Rule](https://junit.org/junit4/javadoc/4.12/org/junit/rules/ExpectedException.html) that expects ArrayIndexOutOfBoundsException, then make the 11th call that should fail.

sabeechen commented on src/main/java/io/wgj/ctci/Stack.java:
```diff
@@ -0,0 +1,21 @@
+    int position;
+
+    public Stack(){
+        stack = new int[10];
```
Consider:
* Using generics for the stack type instead of just using integer.
* Making the maximum stack size passed in to the constructor (instead of hard-coded).

If the max-stack size *should* be hard-coded (not sure what your req's are for this), then I'd recommend making it a constant that can be referenced and used from the tests too.

wgj replied:
Fixed in cd6dc83cc0daf94827e214ba2a7c26f33c9aef00

sabeechen commented on src/test/java/io/wgj/ctci/StackTest.java:
```diff
@@ -0,0 +1,50 @@
+    @Test
+    public void testStackMultipleInts() {
+        ArrayList<Integer> expected = new ArrayList<>();
+        for (int i = 2; i < 10; i = i + 2) {
```
Pedantic: When I look at a test, I expect everything to be straightoward and reasonable.  When I look at this, my first thought is "why does this iterate from 2, 4, ...., 18 instead of 0, 1, ...., 9"?  IMHO things like this make it harder to understand what a test is supposed to be doing because you have to think twice about the intention.

That is, assuming there isn't *actually* a reason.  If thats the case it should have a comment.

wgj replied:
Broke into two tests, one for natural sequences and one for random number, which I think better shows my intent, such that an obvious sequence of 1..10 wouldn't hide a bug.

sabeechen commented on src/test/java/io/wgj/ctci/StackTest.java:
```diff
@@ -0,0 +1,50 @@
+
+import java.util.ArrayList;
+import java.util.Collections;
+
```
Recommend adding a test that has a combination of interleaved push() and pop() calls to ensure you're tracking position correctly.

wgj replied:
Fixed in cd6dc83cc0daf94827e214ba2a7c26f33c9aef00

sabeechen reviewed (commented):
Not sure how strict you wanted me to be as a reviewer, so I turned the scrutiny up to 11.

wgj committed:
Reformat code

wgj committed:
Chain methods for readability

wgj committed:
Add random integer test

testStackMultipleInts() and testStackRandomInts shows intent better, such that
the obvious sequence of 1..10 wouldn't hide a bug.

sabeechen commented on src/test/java/io/wgj/ctci/StackTest.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Test
+    public void testStackRandomInts() {
+        Random r = new Random();
```
This is more of a fuzzer than a test.  If it did find a bug, how would you be able to determine the seed that reproduces the problem?  The best tests do *exactly* the same thing every time.  They can have randomness, but it better be there for a good reason.  As-is this test seems basically identical to testStackMultipleInts except it could be more flaky because of the randomness.

wgj replied:
Fixed in cd6dc83cc0daf94827e214ba2a7c26f33c9aef00

wgj committed:
Use generics, remove unneeded tests, add dynamic array

Also adds a class description, quality of life methods, and associated tests

wgj commented:
@sabeechen Ready for review

sabeechen commented on src/main/java/io/wgj/ctci/Stack.java:
```diff
@@ -1,21 +1,80 @@
+    }
+
+    public Stack(int size) {
+        stack = new Object[size];
```
Setting this before checking constraints on size means your array could be bigger than necessary.  That being said, why have a maximum size at all?  Is it part of the exercise?  Otherwise it seem like it just makes this class less usable and more complicated.

wgj replied:
> Setting this before checking constraints on size means your array could be bigger than necessary. 

Not sure what you mean. In the case where the caller creates a `Stack` without using it, an array size of 10 seems like small enough waste to not be terrible mistake. In the case where the caller sets an initial capacity, and they don't use it seems like a mistake, but I think that's the responsibility of the caller. Guarding against that seems like it would be at the expense of responsible users, wanting to avoid unnecessary growths.

> why have a maximum size at all?

To prevent oom errors. Granted `1024` is too low, but it's gotta be _something_ and it's a toy program. I wanted to promote responsible array use, but I'd imagine setting the maximum size as `Integer.MAX_VALUE` would better show that intent.

> Is it part of the exercise?

No.

> Otherwise it seem like it just makes this class less usable and more complicated.

Agreed. The goal isn't generally usability by another programmer, but familiarity with the stack and dynamic array data structure. With that in mind, I chose to not use unbounded arrays, and to not short change myself on the last growth operation. I thought it was more "right" to solve for those constraints than to pass the buck to the calling programmer.

wgj replied:
Per our conversation, it's generally the caller's responsibility to prevent overflow of the data structure.

sabeechen commented on src/main/java/io/wgj/ctci/Stack.java:
```diff
@@ -1,21 +1,80 @@
+    }
+
+    private void grow() {
+        int newSize = size << 2;
```
I think maybe you meant to multiply by 2?  "<< 2" effectively means "multiply by 4", which is more than necessary to get amortized linear complexity.  It will still work fine, but in practice will lean more heavily on memory usage.

Also you might disagree, but imho the only reason to use bit shifting in java in 2020 is to let people know you liked computers before nerds were cool.  It is unarguably more performant but it just doesn't matter.  It also handles overflows by silently failing (it is NOT multiplication), which might be bad.

wgj replied:
Thought it was a cool way to grow. I stole `<<2` from OpenJDK's implementation of `ArrayList`, which being a major implementation of the API, can obviously benefit from the performance savings.

sabeechen commented on src/main/java/io/wgj/ctci/Stack.java:
```diff
@@ -1,21 +1,80 @@
+    }
+
+    public int maxSize() {
+        return MAX_ARRAY_SIZE;
```
Just make MAX_ARRAY_SIZE public, this doesn't need to be a method.  You're using capacity() to do what I'd previously hoped maxSize() would do (which is fine).

wgj replied:
This is a problem I have with learning Java. I like your suggestion, and even used it for a while, but Sun by way of Oracle makes me think accessing members directly without a getter is worse than killing babies. Could you expand on that?

sabeechen commented on src/main/java/io/wgj/ctci/Stack.java:
```diff
@@ -1,21 +1,80 @@
+        }
+        Object[] newStack = new Object[newSize];
+        size = newSize;
+        stack = newStack;
```
Doesn't this need to copy the old array into the new one?  I'm not sure how this is passing the tests.
[A useful method.](https://docs.oracle.com/javase/7/docs/api/java/util/Arrays.html#copyOf(T[],%20int))

wgj replied:
Huuuuuge miss. Thank you.

wgj replied:
There (previously) wasn't a test for the contents after a growth of the inner array.

sabeechen commented on src/test/java/io/wgj/ctci/StackTest.java:
```diff
@@ -2,49 +2,121 @@
+        s.push(3);
+        int newCapacity = s.capacity();
+        assertTrue(startingCapacity < newCapacity);
+
```
Extra newline

sabeechen commented on src/test/java/io/wgj/ctci/StackTest.java:
```diff
@@ -2,49 +2,121 @@
+    @Test
+    public void testGrowth() {
+        Stack<Integer> s = new Stack<>(1);
+        int startingCapacity = s.capacity();
```
You know what the capacity should be at this point, 1, so I'd just assertEquals(1, s.capacity())

wgj replied:
Since the test's stated goal is the internal growth function, what's the idea on assertions that might be orthogonal? Should a sanity check for initial capacity be in a separate test? Should a dependant states be tested along with the ostensible test? Both?

sabeechen commented on src/main/java/io/wgj/ctci/Stack.java:
```diff
@@ -1,21 +1,80 @@
+    static final int MAX_ARRAY_SIZE = 1024;
+    static final int DEFAULT_CAPACITY = 10;
+    Object[] stack;
+    int size;
```
Looking at your implementation, "size" should actually be called capacity and "position" should be called size.

sabeechen commented on src/test/java/io/wgj/ctci/StackTest.java:
```diff
@@ -2,49 +2,121 @@
+        s.push(2);
+        s.push(3);
+        int newCapacity = s.capacity();
+        assertTrue(startingCapacity < newCapacity);
```
Same here, assertEquals(4, s.capacity())  (at least I think its 4)

wgj replied:
I'd think that using the starting variable prevents that off-by-one arithmetic that humans tend to do, or is it more important to not mask a test's workings by being trying to be clever?

sabeechen commented on src/test/java/io/wgj/ctci/StackTest.java:
```diff
@@ -2,49 +2,121 @@
-            s.push(i);
+    @Test
+    public void testMaxArraySize() {
+        Stack<Integer> s = new Stack<>(1024); // Stack.MAX_ARRAY_SIZE
```
Make MAX_ARRAY_SIZE public, and use it instead of the constant?

wgj replied:
This is a problem I have with learning Java. I like your suggestion, and even used it for a while, but Sun by way of Oracle makes me think accessing members directly without a getter is worse than killing babies. Could you expand on that?

sabeechen commented on src/test/java/io/wgj/ctci/StackTest.java:
```diff
@@ -2,49 +2,121 @@
+    }
+
+    @Test
+    public void TestPeek() {
```
This test and the ones below it have the wrong casing (eg Test vs test)

sabeechen commented on src/test/java/io/wgj/ctci/StackTest.java:
```diff
@@ -2,49 +2,121 @@
+    public void TestPeek() {
+        Stack<Integer> s = new Stack<>();
+        s.push(0);
+        assertEquals(0, (int) s.peek());
```
I'd make these 3 tests push and pop at least 4 times to make sure they update correctly after the array resizes.

wgj replied:
New test `testContentsAfterGrowth()` provides similar coverage.

sabeechen commented on src/main/java/io/wgj/ctci/Stack.java:
```diff
@@ -1,21 +1,80 @@
+        int newSize = size << 2;
+        if (newSize > MAX_ARRAY_SIZE) {
+            int leftOver = MAX_ARRAY_SIZE - size;
+            // Squeeze out the remaining space we'd lose.
```
I might be missing something, but I don't understand how this comment makes sense.

wgj replied:
It should be rewritten, but the comment below shows you understand it as intended.

sabeechen commented on src/main/java/io/wgj/ctci/Stack.java:
```diff
@@ -1,21 +1,80 @@
+            } else {
+                throw new UnsupportedOperationException();
+            }
+        }
```
It took me a couple reads through the block above is doing.  I think its equivalent to having:
```java
int newSize;
if (size == MAX_ARRAY_SIZE) {
  // We're out of room to grow.
  throw new UnsupportedOperationException();
} else if (size < MAX_ARRAY_SIZE / 2) {
  // There is enough space to grow the array
  newSize = size * 2;
} else {
  // Doubling the array would be too big, so cap it at the max size.
  newSize = MAX_ARRAY_SIZE;
}
```

If you can pull it off, having fewer nested scopes is more understandable because of how our brains work.  The cognitive load for following branches within branches is much higher than a flat structure.  For example:
```java
if (a) {
  if (b) {
    // Do thing
  } else {
    // Do other thing
  }
}
```
Is equivalent to but less understandable than:
```java
if (a && b) {
  // do thing
} else {
 // Do other thing
}
```

Some static analysis tools even use nested scopes as a measure of quality, and I can't say that I disagree.  Obviously you can't always simplify like this.  Sometimes you have to rejigger the variables around to do it, butsometimes it just isn't possible to express the logic that way.

wgj replied:
> Some static analysis tools even use nested scopes as a measure of quality.

That's one of the first things we'd show off in a Jtest demo, and ironically I know "McCabe's Cyclomatic Complexity" by name. Shame on me.

sabeechen commented on src/main/java/io/wgj/ctci/Stack.java:
```diff
@@ -1,21 +1,80 @@
+            int leftOver = MAX_ARRAY_SIZE - size;
+            // Squeeze out the remaining space we'd lose.
+            if (leftOver > 0) {
+                newSize = leftOver + size;
```
If I'm reading this right, at this line 
newSize = leftOver + size
but since
leftOver = MAX_ARRAY_SIZE  - size
then,
newSize = (MAX_ARRAY_SIZE  - size) + size = MAX_ARRAY_SIZE
So it would be better to just make this line:
```java
newSize = MAX_ARRAY_SIZE;
```

wgj replied:
Agreed, thank you.

sabeechen commented on src/test/java/io/wgj/ctci/StackTest.java:
```diff
@@ -2,49 +2,121 @@
+        assertEquals(2, (int) s.pop());
+        s.push(4);
+        assertEquals(4, (int) s.pop());
+        assertEquals(1, (int) s.pop());
```
Maybe verify size() == 0 is empty at the end?

wgj committed:
Rename `size` to `capacity` and `position` to `size`

wgj committed:
Add test for array copies

wgj committed:
Copy array contents on growth

wgj committed:
Add test for size on fully pop'd stack

wgj committed:
Multiply by two instead of bitshift two

wgj committed:
Remove MAX_ARRAY_SIZE and improve tests

Refactor grow() to take advantage of unbounded internal array, and use copyOf().
Tests now include more intermixed assertions, and more state changes before
assertions.
Remove tests associated with MAX_ARRAY_SIZE.
